libcudacxx/include/cuda/std/__cccl/attributes.h

```diff
@@ -115,4 +115,19 @@
 #  define _CCCL_RESTRICT __restrict__
 #endif // ^^^ !_CCCL_COMPILER(MSVC) ^^^
 
+// prefer compiler specific way of defining deprecated because of compiler bugs in old GCC and Clang
+#if _CCCL_COMPILER(MSVC)
+#  define _CCCL_DEPRECATED               __declspec(deprecated)
+#  define _CCCL_DEPRECATED_BECAUSE(_MSG) __declspec(deprecated(_MSG))
+#elif _CCCL_COMPILER(GCC) || _CCCL_COMPILER(CLANG) || _CCCL_COMPILER(NVHPC) || _CCCL_COMPILER(ICC)
+#  define _CCCL_DEPRECATED               __attribute__((__deprecated__))
+#  define _CCCL_DEPRECATED_BECAUSE(_MSG) __attribute__((__deprecated__(_MSG)))
+#elif _CCCL_STD_VER >= 2014
+#  define _CCCL_DEPRECATED               [[deprecated]]
+#  define _CCCL_DEPRECATED_BECAUSE(_MSG) [[deprecated(_MSG)]]
```

Comment:
Important: the change of order was pushed after my approval and worsens the status-quo. Was this required because of compilation failures?

I want to get rid of vendor specific attributes as soon as possible, so we should know what compiler is blocking us here.

Reply:
Appologies for the push after approval.

Yes, the compilation fails for older GCC and Clang when combining multiple attributes. See the example https://godbolt.org/z/sarjePhPM

I will find out the exact versions and set the vendor's specific attributes only for them. Is this an acceptable solution?

Reply:
If this affects a lot of compilers, I am also fine with taking your suggestion of prefering vendor attributes. But we should know when we can kick those out, so please add a comment about that. Thx!

Reply:
> Appologies for the push after approval.

We don't have a rule that forbids anybody doing that, and we actually frequently have to push fixes discovered by the CI after approval. So nothing wrong here :) It just happened to change how I feel about the PR.

Reply:
I've come up with a better solution. First, the macro is defined to vendor's specific attributes. However, if C++14 is available I use `[[deprecated]]` instead unless a defected version of GCC or Clang is used.

Hopefully this is an acceptable and functional version :)

Reply:
Works for me! Thx for taking the time to find a better solution.